be/src/formats/parquet/group_reader.cpp

```diff
@@ -101,47 +91,80 @@ Status GroupReader::get_next(ChunkPtr* chunk, size_t* row_count) {
+    size_t hit_count =
+            filtered_chunk_size >= 0 ? filtered_chunk_size : SIMD::count_nonzero(chunk_filter.data(), rows_read);
+
     if (has_filter) {
-        size_t hit_count = chunk_size >= 0 ? chunk_size : SIMD::count_nonzero(chunk_filter.data(), count);
         if (hit_count == 0) {
             active_chunk->set_num_rows(0);
-        } else if (hit_count != count) {
-            active_chunk->filter_range(chunk_filter, 0, count);
+        } else if (hit_count != rows_read) {
+            active_chunk->filter_range(chunk_filter, 0, rows_read);
         }
         active_chunk->check_or_die();
     }
 
+    Status status = Status::OK();
+    if (rows_read < rows_to_read) {
+        status = Status::EndOfFile("");
+    }
+
     size_t active_rows = active_chunk->num_rows();
     if (active_rows > 0 && !_lazy_column_indices.empty()) {
-        ChunkPtr lazy_chunk = _create_read_chunk(_lazy_column_indices);
-        RETURN_IF_ERROR(_lazy_skip_rows(_lazy_column_indices, lazy_chunk, *row_count));
-
         SCOPED_RAW_TIMER(&_param.stats->group_chunk_read_ns);
-        // read data into lazy chunk
-        _column_reader_opts.context->filter = has_filter ? &chunk_filter : nullptr;
-        Status st = _read(_lazy_column_indices, &count, &lazy_chunk);
-        if (!st.ok() && !st.is_end_of_file()) {
-            return st;
+        ChunkPtr lazy_chunk = _create_read_chunk(_lazy_column_indices);
+        // skip previous accumulate rows to skip first
+        if (_previous_rows_to_skip > 0) {
+            ASSIGN_OR_RETURN(size_t rows_skip, _skip(_lazy_column_indices, _previous_rows_to_skip));
+            DCHECK_EQ(rows_skip, _previous_rows_to_skip);
+            _previous_rows_to_skip = 0;
         }
-        if (has_filter) {
-            lazy_chunk->filter_range(chunk_filter, 0, lazy_chunk->num_rows());
+        double selectivity = (double)hit_count / (double)chunk_filter.size();
+        // 0.1 just an YY number
+        if (selectivity >= 0.1) {
+            // using batch way avoid frequently call skip() function, it will introduce lots of virtual function calls
+            ASSIGN_OR_RETURN(size_t tmp_rows_read, _read(_lazy_column_indices, rows_read, &lazy_chunk));
+            DCHECK_EQ(rows_read, tmp_rows_read);
+            if (has_filter) {
+                lazy_chunk->filter_range(chunk_filter, 0, rows_read);
+            }
+        } else {
+            // using incremental way
+            for (size_t index = 0; index < chunk_filter.size();) {
+                bool is_selected = chunk_filter[index++];
+                size_t run = 1;
+                while (index < chunk_filter.size() && chunk_filter[index] == is_selected) {
+                    index++;
+                    run++;
+                }
+                if (is_selected) {
+                    ASSIGN_OR_RETURN(size_t tmp_rows_read, _read(_lazy_column_indices, run, &lazy_chunk));
```

Comment:
Question: does `_read` overwrite `lazy_chunk` from a previous run, or just append it?